engine/lib/phases/phase_drop_match_guards.ml

```diff
@@ -0,0 +1,200 @@
+    end
+
+    [%%inline_defs dmutability]
+
+    let rec dexpr' (span : span) (expr : A.expr') : B.expr' =
+      match expr with
+      | [%inline_arms "dexpr'.*" - Match] -> auto
+      | Match { scrutinee; arms } ->
+          let arms = transform_arms (dexpr scrutinee) (List.rev arms) [] in
+          Match { scrutinee = dexpr scrutinee; arms }
+
+    and transform_arms (scrutinee : B.expr) (remaining : A.arm list)
+        (treated : B.arm list) : B.arm list =
+      match remaining with
+      | [] -> treated
+      | { arm = { arm_pat; body; guard = None }; span } :: remaining ->
+          let new_arm : B.arm =
+            {
+              arm = { arm_pat = dpat arm_pat; body = dexpr body; guard = None };
+              span;
+            }
+          in
+          transform_arms scrutinee remaining (new_arm :: treated)
+      | {
+          arm =
+            {
+              arm_pat;
+              body;
+              guard = Some { guard = IfLet { lhs; rhs; _ }; span = guard_span };
+            };
+          span;
+        }
+        :: remaining ->
+          let result_typ = dty span body.typ in
+          let opt_result_typ : B.ty =
+            TApp
+              {
+                ident = Global_ident.of_name Type Core__option__Option;
+                args = [ GType result_typ ];
+              }
+          in
+          let mk_opt_expr (value : B.expr option) : B.expr =
+            let (name : Concrete_ident.name), args =
+              match value with
+              | Some v -> (Core__option__Option__Some, [ v ])
+              | None -> (Core__option__Option__None, [])
+            in
+            UB.call_Constructor name false args guard_span opt_result_typ
+          in
+
+          let mk_opt_pattern (binding : B.pat option) : B.pat =
+            let (name : Concrete_ident.name), (args : B.field_pat list) =
+              match binding with
+              | Some b ->
+                  ( Core__option__Option__Some,
+                    [ { field = `TupleField (0, 1); pat = b } ] )
+              | None -> (Core__option__Option__None, [])
+            in
+            {
+              p =
+                PConstruct
+                  {
+                    name =
+                      Global_ident.of_name
+                        (Constructor { is_struct = false })
+                        name;
+                    args;
+                    is_record = false;
+                    is_struct = false;
+                  };
+              span = guard_span;
+              typ = opt_result_typ;
+            }
+          in
+
+          let expr_none = mk_opt_expr None in
+          let r : B.expr =
+            {
+              e =
+                Match
+                  {
+                    scrutinee;
+                    arms =
+                      [
+                        UB.make_unguarded_arm (dpat arm_pat)
+                          {
+                            e =
+                              Match
```

Comment:
Can you hoist this inner match to a let binding? I mean in OCaml, not in the generated code.
Then, can you add a comment that explains what is this sub-match exactly, with an example (e.g. a quote of the example you'll have earlier in the code)